src/playlist-maker.js

```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
+  /**
+   * Get the index of the last item in the playlist.
+   *
+   * @return {number}
+   *         The index of the last item in the playlist.
+   */
+  playlist.indexOfLast = () => {
+    return list.length - 1;
+  };
 
-        if (typeof source === 'string') {
-          return indexInSources(list, source);
-        } else if (source.src) {
-          return indexInSources(list, source.src);
-        }
-      }
+  /**
+   * Get the index of the next item in the playlist.
+   *
+   * @return {number}
+   *         The index of the next item in the playlist.
+   */
+  playlist.indexOfNext = () => {
+    const current = playlist.currentItem();
 
+    if (current === -1) {
       return -1;
-    },
+    }
 
-    /**
-     * Plays the first item in the playlist.
-     *
-     * @return {Object|undefined}
-     *         Returns undefined and has no side effects if the list is empty.
-     */
-    first() {
-      if (list.length) {
-        return list[playlist.currentItem(0)];
-      }
+    const lastIndex = playlist.indexOfLast();
 
-      playlist.currentIndex_ = -1;
-    },
+    // When repeating, loop back to the beginning on the last item.
+    if (playlist.repeat_ && current === lastIndex) {
+      return 0;
+    }
 
-    /**
-     * Plays the last item in the playlist.
-     *
-     * @return {Object|undefined}
-     *         Returns undefined and has no side effects if the list is empty.
-     */
-    last() {
-      if (list.length) {
-        return list[playlist.currentItem(list.length - 1)];
-      }
+    // Don't go past the end of the playlist.
+    return Math.min(current + 1, lastIndex);
+  };
 
-      playlist.currentIndex_ = -1;
-    },
+  /**
+   * Get the index of the previous item in the playlist.
+   *
+   * @return {number}
+   *         The index of the previous item in the playlist.
+   */
+  playlist.indexOfPrevious = () => {
+    const current = playlist.currentItem();
 
-    /**
-     * Plays the next item in the playlist.
-     *
-     * @return {Object|undefined}
-     *         Returns undefined and has no side effects if on last item.
-     */
-    next() {
+    if (current === -1) {
+      return -1;
+    }
 
-      let nextIndex;
+    // When repeating, loop back to the end of the playlist.
+    if (playlist.repeat_ && current === 0) {
+      return playlist.indexOfLast();
+    }
 
-      // Repeat
-      if (playlist.repeat_) {
-        nextIndex = playlist.currentIndex_ + 1;
-        if (nextIndex > list.length - 1) {
-          nextIndex = 0;
-        }
+    // Don't go past the beginning of the playlist.
+    return Math.max(current - 1, 0);
+  };
 
-      // Don't go past the end of the playlist.
-      } else {
-        nextIndex = Math.min(playlist.currentIndex_ + 1, list.length - 1);
-      }
+  /**
+   * Plays the first item in the playlist.
+   *
+   * @return {Object|undefined}
+   *         Returns undefined and has no side effects if the list is empty.
+   */
+  playlist.first = () => {
```

Comment:
This name threw me off at first, I expected this function to only be a getter, perhaps it should be `playFirst`?

Reply:
I would think the same thing about the other similar functions

Reply:
I would agree, but I'm not trying to make breaking changes here. 😄